File: filters/mermaid.hs
import Data.ByteString qualified as BS
import Data.ByteString.Base64 qualified as B64
import Data.Maybe
import Data.Text qualified as T
import System.Directory
import System.IO
import System.IO.Temp (writeSystemTempFile)
import System.Process
import Text.Pandoc.JSON

readFileBase64 :: FilePath -> IO T.Text
readFileBase64 filePath = do
  bsFileContent <- BS.readFile filePath
  return (B64.encodeBase64 bsFileContent)

renderMermaid :: Block -> IO Block
renderMermaid cb@(CodeBlock (id, classes, namevals) contents)
  | T.pack "mermaid" `elem` classes = do
      systemTempDir <- getTemporaryDirectory
      let mermaidTempDir = systemTempDir ++ "/mermaid-tmp"
      createDirectoryIfMissing False mermaidTempDir

      let tempFileIn = mermaidTempDir ++ "/mermaid-temp-input.mmd"
      tempFileOut <- writeSystemTempFile "mermaid-temp-output.pdf" ""

      writeFile tempFileIn (T.unpack contents)

      let width = fromMaybe (T.pack "400") (lookup (T.pack "width") namevals)

      -- execute mermaid-cli
      let mermaidCliArgs = ["-i", tempFileIn, "-o", tempFileOut, "--puppeteerConfigFile", "/resources/.puppeteer.json", "--pdfFit", "-w", T.unpack width]
      hPutStrLn stderr ("[Mermaid Filter] Executing mermaid-cli with the following args: " <> unwords mermaidCliArgs)
      readProcess "mmdc" mermaidCliArgs ""

      -- TODO use svg and convert it later using inkscape (librsvg is not stable for converting all mermaid SVGs)
      -- base64Content <- readFileBase64 tempFileOut
      -- let base64Path = T.pack "data:image/svg+xml;base64," <> base64Content

      -- cleanup temp files
      removeFile tempFileIn
      -- removeFile tempFileOut
      -- removeDirectory mermaidTempDir

      --                           \/ TODO allow to pass classes
      return (Para [Image (id, [], []) [Str (T.pack "Placeholder Caption")] (T.pack tempFileOut, T.pack "title")])
renderMermaid x = return x

main :: IO ()
main = toJSONFilter renderMermaid

File: filters/mermaid.lua
PANDOC_VERSION:must_be_at_least("3.0")

local function with_tmp_dir(name, action)
  return pandoc.system.with_temporary_directory(
          name,
          function(tmpdir)
            return pandoc.system.with_working_directory(tmpdir, action)
          end
      )
end

local function file_exists(name)
  local f = io.open(name, "r")
  if f ~= nil then
    f:close()
    return true
  else
    return false
  end
end

local function convert(source_code)
  return with_tmp_dir(
          "mermaid",
          function()
            local mermaid_output = "pandoc_mermaid_output.svg"

            local mermaid_args = { "--output", mermaid_output }

            local puppeteer_config = "/resources/.puppeteer.json"
            if file_exists(puppeteer_config) then
              mermaid_args = { "--puppeteerConfigFile", puppeteer_config, table.unpack(mermaid_args) }
            end

            pandoc.pipe("mmdc", mermaid_args, source_code)

            local read_handle = io.open(mermaid_output, 'r')

            local img
            if read_handle then
              img = read_handle:read('a')
              read_handle:close()
            else
              error("Could not read mermaid output file " .. mermaid_output)
            end

            return img
          end
      )
end

local filetype = "svg"
local mimetype = "image/svg+xml"

-- TODO temporary test variables
local caption_or_alt = "Alt (or Caption?) Text"
local caption = "Caption Text"
local title = "Title Text"
local figure_attrs = {}

function CodeBlock(code_block)
  if not code_block.classes[1] == "mermaid" then
    return nil
  end

  local convert_success, img = pcall(convert, code_block.text)
  if not convert_success then
    -- io.stderr:write(tostring(img or "Mermaid conversion returned no data"))
    io.stderr:write(tostring(img) .. "\n")
    error("Mermaid conversion failed")
  end

  local fname = pandoc.utils.sha1(img .. "." .. filetype)

  pandoc.mediabag.insert(fname, mimetype, img)

  local image_attrs = {
      width = code_block.attributes.width,
      height = code_block.attributes.height,
  }

  local image_obj = pandoc.Image(caption_or_alt, fname, title, image_attrs)

  return pandoc.Figure(image_obj, pandoc.Blocks {}, figure_attrs)
end
